lib/AfricasTalking.rb

```diff
@@ -1,114 +1,111 @@
-		end
-	
-		def airtime
-			return Airtime.new @username, @apikey
-		end
-	
-		def voice
-			return Voice.new @username, @apikey
-		end
-	
-		def token
-			return Token.new @username, @apikey
-		end
-	
-		def application
-			return Application.new @username, @apikey
-		end
-	end
-
-	class AfricasTalkingException < Exception
-		# error handling appear here
-		# def initialize(msg="My default message")
-		#     super
-		# end
-
-	end
-
-
-	private 
-		
-		def sendNormalRequest url_, data_ = nil
-			uri		 	     = URI.parse(url_)
-			http		     = Net::HTTP.new(uri.host, uri.port)
-			http.use_ssl     = true
-			if(data_ != nil)
-				request = Net::HTTP::Post.new(uri.request_uri)
-				request.set_form_data(data_)
-			else
-				request = Net::HTTP::Get.new(uri.request_uri)
-			end
-			request["apikey"] = @apikey
-			request["Accept"] = "application/json"
-			response          = http.request(request)
-
-			if (DEBUG)
-				puts "Full response #{response.body}"
-			end
-
-			@response_code = response.code.to_i
-			return response.body
-		end
-
-		def sendJSONRequest url_, data_, get_request = false
-			uri	       = URI.parse(url_)
-			http         = Net::HTTP.new(uri.host, uri.port)
-			http.use_ssl = true
-			if get_request === true
-				uri.query = URI.encode_www_form(data_)
-				req = Net::HTTP::Get.new(uri.request_uri, 'Content-Type'=>"application/json")
-			else
-				req = Net::HTTP::Post.new(uri.request_uri, 'Content-Type'=>"application/json")
-				req.body = data_.to_json
-			end
-			
-			req["apikey"] = @apikey
-			req["Accept"] = "application/json"
-			response  = http.request(req)
-			if (DEBUG)
-				puts "Full response #{response.body}"
-			end
-
-			@response_code = response.code.to_i
-			return response.body
-		end
-
-		def validateParamsPresence? params, values
-			status =  values.each{ |v|
-				if !params.key?(v)
-					raise AfricasTalkingException, "Please make sure your params has key #{v}"
-				elsif v.empty?
-					raise AfricasTalkingException, "Please make sure your key #{v} is not empty"
-				end
-			}
-			return true
-		end
+  DEBUG = false
+  @username = nil
+  @apikey = nil
+  @response_code = nil
+  # class Initialize
+  class Initialize
+    attr_accessor :username, :apikey
+    def initialize(username, apikey)
+      @username    = username
+      @apikey      = apikey
+    end
+
+    def sms
+      Sms.new @username, @apikey
+    end
+
+    def payments
+      Payments.new @username, @apikey
+    end
+
+    def airtime
+      Airtime.new @username, @apikey
+    end
+
+    def voice
+      Voice.new @username, @apikey
+    end
+
+    def token
+      Token.new @username, @apikey
+    end
+
+    def application
+      Application.new @username, @apikey
+    end
+  end
+
+  class AfricasTalkingException < RuntimeError
+    # error handling appear here
+    # def initialize(msg="My default message")
+    #     super
+    # end
+  end
+
+  private
+
+  def sendNormalRequest(url_, data_ = nil)
```

Comment:
out of curiosity, why do we have these underscore suffixes? What would be wrong with `url` and `data`, without the underscores? :slightly_smiling_face:

Reply:
It works but a wrong choice of variables. I would recommend keyword arguments instead.
```ruby
def send_normal_request( url: , data:)
```

Reply:
@DrBenton 

> out of curiosity, why do we have these underscore suffixes? What would be wrong with `url` and `data`, without the underscores? 🙂

The underscores are used if you want to surpress the `unused block parameters` warning you can get when you use rubocop for example to analyze code style guides. Maybe that's the history behind the choice of these variable names, but I guess it doesn't make much sense anymore since the variables are used.